arboretum/common/utils.py

```diff
@@ -34,3 +38,25 @@ def parse_seconds(seconds):
             formatted.append(f'{q} {unit}')
         seconds = r
     return ', '.join(formatted)
+
+
+def run_command(cmd, secrets=None):
+    """Run commands in a system."""
+    if type(cmd) == str:
+        cmd = cmd.split(' ')
```

Comment:
my intention for `cmd` is that it should accept `str` or list of `str` because `subprocess.Popen` accepts `str` (if `shell=True`) or list of `str` (if `shell=False`).  By accepting both types, user of this function does not need to care about the internal implementation.

To clearly show the intention (and apply your comment about `type` and `isinstance`), I plan to modify the code as follows.

```python
    """
    Execute system command.

    :param cmd: a space-separated string or a list of string
    :param secrets: a text which should be masked in log text.
    :returns: standard output of the command.
    """
    if isinstance(cmd, str):
        cmd = cmd.split(' ')
    elif not isinstance(cmd, list):
        raise TypeError('given command line was neither '
                        f'a space-separated string nor list of string: {cmd}')
```

Reply:
> my intention for cmd is that it should accept str or list of str because subprocess.Popen accepts str (if shell=True) or list of str (if shell=False)

But that's not what's happening in `run_command` currently.  There's no logic to optionally set the `shell=True` parameter for `Popen` and it will always be acting on a list of arguments anyway based on the previous and proposed versions of `run_command` because  if `cmd` comes in as a string, you turn it into a list of arguments.

1.  If you truly want `run_command` to dynamically either execute `cmd` as a string or `cmd` as a list of arguments then `if isinstance(cmd, str):` is true you should apply `shell=True` as a parameter to the `Popen` call.
2.  However, if you want to continue to break `cmd` into arguments then you should use `shlex.split` rather than `str.split`.

Either solution (1) or (2) will achieve:

> ...By accepting both types, user of this function does not need to care about the internal implementation.

A few other things...

- I think you should look into using [subprocess.run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) rather than directly using `Popen`.
- You can then leverage https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError for error handling
- It's not necessary to raise a TypeError is cmd is not a string or list.

Reply:
According to your comments and the document of [subprocess](https://docs.python.org/3.7/library/subprocess.html), I plan to change the implementation as follows.  I really appreciate your comments and suggestions on this, thanks.

- use `subprocess.run()` instead of `subprocess.Popen()`
- assume only `list[str]` for `cmd` to leverage functionality (quoting, etc.) of `subprocess.run()`
- delegate error handling to `subprocess.run()`, and use `subprocess.CalledProcessError` instead of custom exception class `CommandExecutionError`
    - to do so, externalize the masking feature to another function
- other changes (accept further parameters and return stderr in addition to stdout for future use)

```python
def mask_secrets(text, secrets):
    """
    Replace secret words in a text with `***`.

    :param str text: a string which may contain secret words.
    :param list[str] secrets: secret word list.
    :returns: masked text.
    """
    for s in secrets:
        text = text.replace(s, '***')
    return text


def run_command(cmd, input_text=None, timeout=None):
    """
    Execute system command.

    This is a wrapper for `subprocess.run()`.

    Example 1: `run_command(['echo', '-n', 'hello'])` returns `('hello','')`.

    Example 2: `run_command(['cat'], input='hello')` returns `('hello','')`.

    Use `subprocess.run()` if other complicated parameters (e.g., encoding)
    should be specified.

    :param list[str] cmd: command line arguments
    :param str input_text: text for standard input of command
    :param int timeout: timeout for command in seconds
    :raises subprocess.CalledProcessError: if the command finishes with
                                           non-zero returncode.
    :raises subprocess.TimeoutExpires: if timeout expires.
    :raises TypeError: if some of `cmd` element is not a `str`.
    :raises IndexError: if length of `cmd` is zero.
    :returns: a tuple of standard output and standard error of the command.
    """
    cp = subprocess.run(
        cmd,
        input=input_text,
        text=True,
        timeout=timeout,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=True,
        shell=False
    )
    return cp.stdout, cp.stderr
```